File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char>st;

        // t.c = O(n)
        // s.c = O(n)

        for(int i = 0; i < s.length(); i++)
        {
            if(s[i] == '(')
            {
                st.push(s[i]);
            }
            if(s[i] == '[')
            {
                st.push(s[i]);
            }
            if(s[i] == '{')
            {
                st.push(s[i]);
            }

            if(s[i] == ')')
            {
                if(!st.empty() and st.top() == '(')
                {
                    st.pop();
                }
                else
                {
                    return false;
                }
            }

            if(s[i] == ']')
            {
                if(!st.empty() and st.top() == '[')
                {
                    st.pop();
                }
                else
                {
                    return false;
                }
            }
            
            if(s[i] == '}')
            {
                if(!st.empty() and st.top() == '{')
                {
                    st.pop();
                }
                else
                {
                    return false;
                }
            }
        }
        
        if(st.empty())
        {
            return true;
        }

        return false;
        
    }
};

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> st = new Stack<Character>();

        for(char ch : s.toCharArray())
        {
            if(st.isEmpty())
            {
                st.push(ch);
            }
            else
            {
                char topval = st.peek();
                if(ch == ')')
                {
                    if(topval == '(')
                    {
                        st.pop();
                    }
                    else
                    {
                        st.push(ch);
                    }
                }
                else if(ch == ']')
                {
                    if(topval == '[')
                    {
                        st.pop();
                    }
                    else
                    {
                        st.push(ch);
                    }

                }
                else if(ch == '}')
                {
                    if(topval == '{')
                    {
                        st.pop();
                    }
                    else
                    {
                        st.push(ch);
                    }

                }
                else 
                {
                    st.push(ch);
                }
            }
        }

        return st.isEmpty();
        
    }
}